### PR TITLE
[stdlib] Adds macros for option

### PR DESCRIPTION
**Pull request by damirka**
## Description 

Adds macros for option, currently, contains `do` and `destroy` (identical).

## Test plan 

Features tests.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: crates/sui-framework/packages/move-stdlib/sources/option.move [hunk 1]
```diff
@@ -141,4 +141,100 @@ module std::option {
...
+        let o = $o;
+        o.do!($f);
+    }
+
+    /// Destroy `Option<T>` and call the closure `f` on the value inside if it holds one.
+    public macro fun do<$T>($o: Option<$T>, $f: |$T|) {
+        let o = $o;
+        if (o.is_some()) {
+            $f(o.destroy_some());
+        }
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/sources/option.move:158]: style nit
```suggestion
        if (o.is_some()) $f(o.destroy_some())
```

File: crates/sui-framework/packages/move-stdlib/sources/option.move [hunk 2]
```diff
@@ -141,4 +141,100 @@ module std::option {
...
+    }
+
+    /// Execute a closure on the value inside `t` if it holds one.
+    public macro fun do_ref<$T>($o: &Option<$T>, $f: |&$T|) {
+        let o = $o;
+        if (o.is_some()) {
+            $f(o.borrow());
+        }
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/sources/option.move:166]:
```suggestion
        if (o.is_some()) $f(o.borrow())
```

File: crates/sui-framework/packages/move-stdlib/sources/option.move [hunk 3]
```diff
@@ -141,4 +141,100 @@ module std::option {
...
+    }
+
+    /// Execute a closure on the mutable reference to the value inside `t` if it holds one.
+    public macro fun do_mut<$T>($o: &mut Option<$T>, $f: |&mut $T|) {
+        let o = $o;
+        if (o.is_some()) $f(o.borrow_mut());
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/sources/option.move:172]:
```suggestion
        if (o.is_some()) $f(o.borrow_mut())
```

File: crates/sui-framework/packages/move-stdlib/sources/option.move [hunk 4]
```diff
@@ -141,4 +141,100 @@ module std::option {
...
+    }
+
+    /// Select the first `Some` value from the two options, or `None` if both are `None`.
+    /// Equivalent to Rust's `a.or(b)`.
+    public macro fun or<$T>($o: Option<$T>, $default: Option<$T>): Option<$T> {
+        let o = $o;
+        if (o.is_some()) o
+        else $default
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/sources/option.move:180]: 🔥

File: crates/sui-framework/packages/move-stdlib/sources/option.move [hunk 5]
```diff
@@ -141,4 +141,100 @@ module std::option {
...
+    public macro fun map_ref<$T, $U>($o: &Option<$T>, $f: |&$T| -> $U): Option<$U> {
+        let o = $o;
+        if (o.is_some()) some($f(o.borrow()))
+        else none()
+    }
+
+    /// Return `None` if the value is `None`, otherwise return `Option<T>` if the predicate `f` returns true.
+    public macro fun filter<$T: drop>($o: Option<$T>, $f: |&$T| -> bool): Option<$T> {
+        let o = $o;
+        if (o.is_some() && $f(o.borrow())) o
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/sources/option.move:219]: Beautiful, love how this one turned out

File: crates/sui-framework/packages/move-stdlib/tests/option_tests.move [hunk 6]
```diff
@@ -169,4 +169,52 @@ module std::option_tests {
...
+
+    // === Macros ===
+
+    #[test]
+    fun do_destroy() {
+        let mut counter = 0;
+        option::some(5).destroy!(|x| counter = x);
+        option::some(10).do!(|x| counter = counter + x);
+
+        assert!(counter == 15);
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/tests/option_tests.move:181]: I snuck in `std::unit_test::assert_eq` a while back if you want to give it a shot. This is fine though

**Review by tnowacki — APPROVED** [review]
Looks amazing! Thank you so much!

Just some style nits. 
Also I would ask that we make sure @tzakian gives this a once over before we land, just since that he helped come up with a lot of the names for the macros here

**Review by tzakian — APPROVED** [review]
Looks amazing! 🚀

**Pull request merged**